src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -874,6 +907,142 @@ void ImageReplayer<I>::shut_down_journal_replay(bool cancel_ops)
 }
 
 template <typename I>
+void ImageReplayer<I>::update_mirror_image_status(bool from_timer, bool reschedule)
+{
+  cls::rbd::MirrorImageStatus status;
+
+  {
+    Mutex::Locker locker(m_lock);
+
+    switch (m_state) {
+    case STATE_UNINITIALIZED:
+      // TODO: a better way to detect error state?
+      if (!m_state_desc.empty()) {
+	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_ERROR;
+	status.description = m_state_desc;
+      } else {
+	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_UNKNOWN;
+      }
+      break;
+    case STATE_STARTING:
+      // TODO: a better way to detect syncing state.
+      if (!m_asok_hook) {
+	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_SYNCING;
+	status.description = m_state_desc.empty() ? "syncing" : m_state_desc;
+      } else {
+	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STARTING_REPLAY;
+	status.description = "starting replay";
+      }
+      break;
+    case STATE_REPLAYING:
+      status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_REPLAYING;
+      status.description = get_replay_status_description();
+      break;
+    case STATE_STOPPING:
+      if (m_local_image_ctx) {
+	status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPING_REPLAY;
+	status.description = "stopping replay";
+	break;
+      }
+      /* FALLTHROUGH */
+    case STATE_STOPPED:
+      status.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPED;
+      status.description = m_state_desc.empty() ? "stopped" : m_state_desc;
+      break;
+    default:
+      assert(!"invalid state");
+    }
+  }
+
+  dout(20) << "status=" << status << dendl;
+
+  librados::ObjectWriteOperation op;
+  librbd::cls_client::mirror_image_status_set(&op, m_global_image_id, status);
+  librados::AioCompletion *comp = librados::Rados::aio_create_completion();
+  int r = m_local_ioctx.aio_operate(RBD_MIRRORING, comp, &op);
+  assert(r == 0);
+  comp->release();
+
+  reschedule_update_status_task(from_timer, reschedule);
+}
+
+template <typename I>
+void ImageReplayer<I>::reschedule_update_status_task(bool from_timer,
+						     bool reschedule)
+{
+  Mutex::Locker locker(m_threads->timer_lock);
+
+  if (m_update_status_task) {
+    if (!from_timer) {
+      m_threads->timer->cancel_event(m_update_status_task);
+    }
+    m_update_status_task = nullptr;
+  }
+
+  if (!reschedule) {
+    return;
+  }
+
+  m_update_status_task = new FunctionContext(
+    [this](int r) {
+      start_update_status_task();
+    });
+
+  // TODO: make status update interval configurable.
+  m_threads->timer->add_event_after(30, m_update_status_task);
+}
+
+template <typename I>
+void ImageReplayer<I>::start_update_status_task()
+{
+  FunctionContext *ctx = new FunctionContext(
+    [this](int r) {
```

Comment:
If  you take ```m_threads->timer_lock``` here, you could set ```m_update_status_task = nullptr;``` here and then eliminate the ```bool from_timer``` argument to ```reschedule_update_status_task``` and ```update_mirror_image_status```.